faunadb/client.go

```diff
@@ -235,6 +262,114 @@ func (client *FaunaClient) NewWithObserver(observer ObserverCallback) *FaunaClie
 	return client.newClient(client.basicAuth, observer)
 }
 
+// Stream creates a stream subscription to the result of the given expression.
+//
+// The subscription returned by this method does not issue any requests until
+// the subscription's Start method is called. Make sure to
+// subscribe to the events of interest, otherwise the received events are simply
+// ignored.
+func (client *FaunaClient) Stream(query Expr, config ...StreamConfig) StreamSubscription {
+	return newSubscription(client, query, config...)
+}
+
+func (client *FaunaClient) startStream(subscription *StreamSubscription) (err error) {
+	var payload []byte
+	var response faunaResponse
+	if subscription.Status() != StreamConnIdle {
+		err = errors.New("stream subscription already started")
+		return
+	}
+
+	startTime := time.Now()
+	payload, err = client.prepareRequestBody(subscription.query)
+	if err != nil {
+		err = errors.New("test error")
+	}
+	body := ioutil.NopCloser(bytes.NewReader(payload))
+
+	var endpoint strings.Builder
+	endpoint.WriteString(client.streamEndpoint)
+
+	if len(subscription.config.Fields) > 0 {
+		endpoint.WriteString("?fields=")
+		count := len(subscription.config.Fields)
+		for i := 0; i < count; i++ {
+			endpoint.WriteString(string(subscription.config.Fields[i]))
+			if i <= count-1 {
+				endpoint.WriteString(",")
+			}
+		}
+	}
+
+	subscription.status.Set(StreamConnOpening)
+	response, err = client.performRequest(body, endpoint.String(), true, nil)
+	if err != nil {
+		err = errors.New("test error")
+	}
+
+	httpResponse := response.response
+	_ = client.storeLastTxnTime(httpResponse.Header)
+	if err = checkForResponseErrors(httpResponse); err != nil {
+
+		subscription.status.Set(StreamConnError)
+
+		httpResponse.Body.Close()
+		response.cncl()
+		return
+	}
+
+	go func() {
+		//subscription.Wg.Add(1)
+		subscription.status.Set(StreamConnActive)
+
+		defer httpResponse.Body.Close()
+		defer response.cncl()
+
+		for {
+			var obj Obj
+
+			select {
+			case <-subscription.closingMessage:
+				subscription.status.Set(StreamConnClosed)
+				close(subscription.eventsMessages)
+				break
+			case <-subscription.getNext:
+				if val, err := client.parseResponse(httpResponse, subscription.query, true, startTime); err != nil {
+
+					if err == io.EOF {
+						subscription.status.Set(StreamConnClosed)
+						break
+					}
+					subscription.eventsMessages <- ErrorEvent{
+						txn: startTime.Unix(),
```

Comment:
I think we should differentiate fauna errors from driver errors so we don't introduce synthetic transaction timestamps.

Reply:
Do I need to delete txn param from ErrorEvent or just find a way to obtain error timestamp from fauna error ?